rfcs/45-stageless.md

```diff
@@ -0,0 +1,871 @@
+  - Some computational overhead, as we must create, evaluate and then deduplicate if-needed ordering constraints over each subgraph.
+- **Model 2:** Do nothing, report execution order ambiguities and allow the user to resolve this by adding additional constraints.
+  - Reduces risk of accidentally creating unnecessary constraints.
+  - More explicit, and thus more verbose.
+  - Relies on users checking and resolving execution order ambiguities.
+
+But which behavior is correct?
+Consider the following concrete case:
+
+We have five relevant systems:
+
+1. `determine_player_movement`: reads `Input`, writes `PlayerIntent`
+2. `collision_detection`: reads `Position` and `Velocity`, writes `Events<Collisions>`
+3. `collision_handling`: reads `Events<Collision>`, writes `Velocity`
+4. `apply_player_movement:` reads `PlayerIntent`, writes `Velocity`
+5. `apply_velocity:` reads `Velocity`, writes `Position`
+
+Suppose our user specifies the following if-needed ordering constraints:
+
+1. `determine_player_movement` is before `collision_detection`
+   1. Spurious: these systems are compatible.
+2. `collision_detection` is before `collision_handling`
+   1. Real: these systems conflict on `Events<Collisions>`
+3. `collision_handling` is before `apply_velocity`
+   1. Real: these systems conflict on `Velocity`
+4. `apply_player_movement` is after `collision_handling`
+   1. Spurious: these systems are compatible.
+5. `apply_player_movement` is before `apply_velocity`
+   1. Real: these systems conflict on `Velocity`
+
+The problem here is that there's no direct link between `determine_player_movement` and `apply_player_movement`: they are ambiguous under Model 2, and the player may move according to stale input!
+
+Under the transitive Model 1, additional if-needed constraints are created between `determine_player_movement` and `apply_player_movement` via the constraint chain (1, 2, 4).
+Then, the inferred constraint between `determine_player_movement` and `apply_player_movement` is converted into a strict ordering constraint, as the systems are incompatible.
+Finally, all redundant strict ordering constraints (such as between `collision_detection` and `apply_velocity`) are discarded for performance reasons.
+
+Under Model 2, the behavior is much simpler: `determine_player_movement` and `apply_player_movement` are ambiguous.
+
+Model 1 preserves the "logical" behavior, but in a very implicit fashion that is hard to reason about.
+Model 2 simply causes the user's logic to break.
+
+By adding powerful (and prominent) tools for detecting spurious ordering constraints and execution order ambiguities, we can follow the expected (and convenient) transitive property while helping users quash strange, fragile bugs as soon as they're introduced.
+
+## Unresolved questions
+
+- Should on-update systems of stages be handled using run criteria?
+  - Currently, all of the update systems of the next state will be handled on the same frame

```

Comment:
I think "stages" is typo for "states" here.

Maybe clarify the inner bullet point. I'm genuinely not sure if what you've described is the current behavior, but the question seems to be asking if "on update" should behave like "on enter/exit" and have its own schedule.